### ISSUE TITLE
Use Go 1.20 `errors.Join` instead of `go.uber.org/multierr`

### ISSUE THREAD

**Issue by mx-psi** [open-telemetry/opentelemetry-collector#8210]
Go 1.20 includes new builtin [error joining support](https://pkg.go.dev/errors#Join) that can replace the usage of `go.uber.org/multierr`. We should switch to using these and ban the usage of `go.uber.org/multierr` via `depguard` (same as [this](https://github.com/open-telemetry/opentelemetry-collector/blob/9d7deb4098e51207ba6e3a95d76c0d3b152a8d11/.golangci.yml#L106-L107)).

This is **blocked by** #8207

**Comment by shivanshu1333** [comment]
@mx-psi kindly assign this one as well

**Comment by bogdandrutu** [comment]
@mx-psi can you document why we stopped this effort until better understand the performance issue.

**Comment by mx-psi** [comment]
I removed the `good first issue` and `easy` labels and added `on hold` instead.

The 'easy' way to change the code to use `errors.Join` means creating a slice of errors, which in the common case will all be nil, and then joining that slice. `multierr.Append` instead avoids this by checking if the error is nil before adding to the internal slice, which is more performant.

We can still use `errors.Join` and avoid this by checking that the errors are not nil before appending to the slice, but it leads to more verbose code.

**Comment by dmathieu** [comment]
Note that `errors.Join` does check if the error is nil, and skips adding those into the internal slice.
https://github.com/golang/go/blob/a9922d096f0de877fba68739b35367d4c25f6ecb/src/errors/join.go#L33-L35
So I'm not sure we do need to check that errors aren't nil before joining them.

I just proposed this additional patch to the Go team which will optimize the case where a single non-nil error is provided, and it can be returned as-is.
https://github.com/golang/go/pull/70770

**Comment by mx-psi** [comment]
@dmathieu The patch looks great, thanks for doing this :) 

I think the problem I was pointing at above is not addressed by existing code. The problem is the difference between these two:
```go
plainJoin := errors.Join(e1, e2, e3)
nestedJoin := errors.Join(errors.Join(e1, e2), e3)

plainAppend := multierr.Append(e1, e2, e3)
nestedAppend := multierr.Append(multierr.Append(e1, e2), e3)
```

I believe `plainJoin != nestedJoin` but `plainAppend != nestedAppend`. The nested version is pretty common. For example see this code: https://github.com/open-telemetry/opentelemetry-collector/blob/808fb7c26039a94965718859ed3c7e53a0daa1ba/service/service.go#L188-L198

Only `multierr.Append` has [this optimization](https://github.com/uber-go/multierr/blob/de75ae527b39a27afcb50a84427ec7b84021d5f4/error.go#L444-L448). This is mentioned in golang/go/issues/65676, and was apparently rejected, which means you have to handle the slice creation yourself if you don't want to have this nested structure (which shouldn't matter most of the time but is pressumably less efficient).

**Comment by dmathieu** [comment]
Yes, `errors.Join` will keep nesting new structs for each call of `errors.Join()`.
I think Go could possibly still optimize that by detecting that the first error is an `*joinError`, and using it rather than a new struct.
